include/envoy/matcher/matcher.h

```diff
@@ -0,0 +1,174 @@
+  // matching requirements). If the match couldn't be completed, {false, {}} will be returned.
+  // If a match result was determined, {true, action} will be returned. If a match result was
+  // determined to be no match, {true, {}} will be returned.
+  virtual MatchResult match(const DataType& matching_data) PURE;
+};
+
+// InputMatcher provides the interface for determining whether an input value matches.
+class InputMatcher {
+public:
+  virtual ~InputMatcher() = default;
+
+  /**
+   * whether the provided input is a match.
+   * @param absl::optional<absl::string_view> the value to match on. Will be absl::nullopt if the
+   * lookup failed.
+   */
+  virtual bool match(absl::optional<absl::string_view> input) PURE;
+};
+
+using InputMatcherPtr = std::unique_ptr<InputMatcher>;
+
+/**
+ * Factory for registering custom input matchers.
+ */
+class InputMatcherFactory : public Config::TypedFactory {
+public:
+  virtual InputMatcherPtr createInputMatcher(const Protobuf::Message& config) PURE;
+
+  std::string category() const override { return "envoy.matching.input_matcher"; }
+};
+
+struct DataInputGetResult {
+  enum class DataAvailability {
+    // The data is not yet available.
+    NotAvailable,
+    // Some data is available, but more might arrive.
+    MoreDataAvailable,
+    // All the data is available.
+    AllDataAvailable
+  };
+
+  DataAvailability data_availability_;
+  // The resulting data. This will be absl::nullopt if we don't have sufficient data available (as
+  // per data_availability_) or because no value was extracted. For example, consider a DataInput
+  // which attempts to look a key up in the map: if we don't have access to the map yet, we return
+  // absl::nullopt with NotAvailable. If we have the entire map, but the key doesn't exist in the
+  // map, we return absl::nullopt with AllDataAvailable.
+  absl::optional<absl::string_view> data_;
+
+  friend std::ostream& operator<<(std::ostream& out, const DataInputGetResult& result) {
+    out << "data input: " << (result.data_ ? result.data_.value() : "n/a");
+    switch (result.data_availability_) {
+    case DataInputGetResult::DataAvailability::NotAvailable:
+      out << " (not available)";
+      break;
+    case DataInputGetResult::DataAvailability::MoreDataAvailable:
+      out << " (more data available)";
+      break;
+    case DataInputGetResult::DataAvailability::AllDataAvailable:;
+    }
+    return out;
+  }
+};
+
+/**
+ * Interface for types providing a way to extract a string from the DataType to perform matching on.
+ */
+template <class DataType> class DataInput {
+public:
+  virtual ~DataInput() = default;
+
+  virtual DataInputGetResult get(const DataType& data) PURE;
+};
+
+template <class DataType> using DataInputPtr = std::unique_ptr<DataInput<DataType>>;
+
+/**
+ * Factory for data inputs.
+ */
+template <class DataType> class DataInputFactory : public Config::TypedFactory {
```

Comment:
Do we expect to have different set of extensions for different `DataType` types?
Should `category` reflect the type of exact `DatatType` to avoid accidental registration under wrong specialization?

Reply:
Good point, I think we probably would want to have each type of DataInputFactory have its own category